gordon/record_checker.py

```diff
@@ -0,0 +1,112 @@
+import asyncio
+import logging
+import time
+
+from async_dns import types
+from async_dns.resolver import ProxyResolver
+
+
+class RecordChecker(object):
+    QUERYING_THRESHOLD = 60
+
+    def __init__(self, dns_ip):
+        """Setup RecordChecker object
+
+        Args:
+            dns_ip: DNS ip server to query
+        """
+        self._dns_ip = dns_ip
+        self._resolver = ProxyResolver()
+        self._resolver.set_proxies([self._dns_ip])
+
+    def _extract_record_data(self, record):
+        """Extract record data from a record dict
+
+        Args:
+            record (dict): DNS record as a dict
+        """
+        return record.values()
+
+    async def check_record(self, record):
+        """Asynchronously queries a DNS server for a given record
+            to check how long it takes the record to become available.
+
+        Args:
+            record (dict): DNS record as a dict
+            with record properties
+
+        """
+        start_time = time.time()
+
+        name, rr_data, r_type, ttl = self._extract_record_data(record)
+        r_type_code = types.get_code(r_type)
+
+        record_found_in_dns_srv = False
+        query_tries = 0
+
+        while not record_found_in_dns_srv and\
+                self.QUERYING_THRESHOLD != query_tries:
+
+            query_tries += 1
+            resolver_res = await self._resolver.query(name, r_type_code)
+            possible_ans = resolver_res.an
+
+            record_found_in_dns_srv = \
+                await self._check_resolver_ans(possible_ans, name,
+                                               rr_data, ttl, r_type_code)
+            await asyncio.sleep(1)
+
+        if not record_found_in_dns_srv:
+            logging.info(
+                f'Sending metric record-checker-failed: {record}')
+        else:
+            final_time = float(time.time() - start_time)
+            logging.info(final_time)
+
+    async def _check_resolver_ans(
+            self, dns_ans_lst, name, rr_data, ttl, r_type_code):
+        """Check if resolver ans is equal to record data.
+
+        Args:
+            dns_ans_lst (list): DNS answer list contains record objects
+            name (str): record name
+            rr_data (list): list of ips for the record
+            ttl (int): record time ot live info
+            r_type_code (int): record type code
+
+        Returns:
+            boolean indicating if DNS ans data is equal to record data
+        """
+        type_filtered_lst = \
+            list(filter(lambda r: r.qtype == r_type_code, dns_ans_lst))
+
+        # check to see that type_filtered_lst has
+        # the same number of records as record rr_data
+        if len(type_filtered_lst) != len(rr_data):
```

Comment:
I'm not sure this check is needed if you are checking below to see if the data matches exactly. Also, this check would be a false positive in a case like this: An old record's data is `[1.1.1.1, 2.2.2.2]` but the new record's data has `[2.2.2.2, 3.3.3.3]`, this check would "pass" just because it's old record and new record has the same length.

Reply:
yes, but its prior test, after that we will check for each ip, so this test covers the case when we get an answer from the dns server for example: 3 record: 1.1.1.1, 2.2.2.2, 3.3.3.3.. and our record has only 2: 1.1.1.1, 2.2.2.2, so the next for loop will return True but this test will return False...